ConsensusSafety.thy

```diff
@@ -43,174 +70,226 @@ fun est :: "message \<Rightarrow> consensus_value"
+lemma extract_without_lambda :
+  "\<forall> params p_set. (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> p \<in> {\<lambda>\<sigma>'. p \<sigma>' | p. p \<in> p_set}. p \<sigma>) \<longrightarrow> (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> p \<in> p_set. p \<sigma>)"
+  by simp
+
+lemma extract_lambda :
+  "\<forall> params p_set. (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> p \<in> {\<lambda>\<sigma>'. p \<sigma>' | p. p \<in> p_set}. p \<sigma>) \<longrightarrow> (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> p \<in> p_set. (\<lambda>\<sigma>'. p \<sigma>') \<sigma>)"
+  by simp
+
+lemma extract_lambda_with_exists :
+  "\<forall> params q_set. (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> p \<in> {\<lambda>\<sigma>'. \<exists> c \<in> \<epsilon> params \<sigma>'. q \<sigma>' | q. q \<in> q_set}. p \<sigma>) \<longrightarrow> (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> q \<in> q_set. (\<lambda>\<sigma>'. \<exists> c \<in> \<epsilon> params \<sigma>'. q \<sigma>') \<sigma>)"
+  apply simp
+
+
+lemma extract_lambda_with_forall :
+  "\<forall> params q_set. (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> p \<in> {\<lambda>\<sigma>'. \<forall> c \<in> \<epsilon> params \<sigma>'. q c | q. q \<in> q_set}. p \<sigma>) \<longrightarrow> (\<exists> \<sigma> \<in> \<Sigma> params. \<forall> q \<in> q_set. (\<lambda>\<sigma>'. \<forall> c \<in> \<epsilon> params \<sigma>'. q c) \<sigma>)"
+  apply simp  
+
+lemma tmp1 :
+  "\<forall> params q_set. \<exists> \<sigma> \<in> \<Sigma> params. 
+    (\<forall> p \<in> {\<lambda>\<sigma>'. \<forall> c \<in> \<epsilon> params \<sigma>'. q c | q. q \<in> q_set}. p \<sigma>
+    \<longrightarrow> (\<forall> c \<in> \<epsilon> params \<sigma>. \<forall> q \<in> q_set. q c))"
+  apply simp
+
+lemma tmp2 :
+  "\<forall> params q_set. \<exists> \<sigma> \<in> \<Sigma> params. 
+    (\<forall> p \<in> {\<lambda>\<sigma>'. \<forall> c \<in> \<epsilon> params \<sigma>'. q c | q. q \<in> q_set}. p \<sigma>
+    \<longrightarrow> (\<exists> c \<in> \<epsilon> params \<sigma>. \<forall> q \<in> q_set. q c))"
+  apply simp
+
+ *)
 lemma naturally_corresponding_consistency :
-  "\<forall> p_set. state_properties_are_consistent w e {naturally_corresponding_state_property w e p | p. p \<in> p_set}
-    \<Longrightarrow> consensus_value_properties_are_consistent p_set"
-  using state_properties_are_consistent_def by auto
+  "\<forall> params q_set. state_properties_are_consistent params {naturally_corresponding_state_property params q | q. q \<in> q_set}
+    \<longrightarrow> consensus_value_properties_are_consistent params q_set"
+  apply auto
+
 
 (* Definition 3.10 *)
-definition consensus_value_property_is_decided :: "threshold \<Rightarrow> weight \<Rightarrow> estimator \<Rightarrow> consensus_value_property \<Rightarrow> state \<Rightarrow> bool"
+fun consensus_value_property_is_decided :: "params \<Rightarrow> (consensus_value_property * state) \<Rightarrow> bool"
```

Comment:
Definition 3.10 causes an error in my environment. 🤔 
![err](https://user-images.githubusercontent.com/1025957/50876876-8aabea00-1412-11e9-863a-5fc44a5c6e82.png)

Reply:
ah yes

Reply:
because the above lemma `naturally_corresponding_consistency` is not proved